docs/typing-indicators.md

```diff
@@ -0,0 +1,146 @@
+input, then we send a request with an `op` of `stop`.  We also send
+"stop" messages when the user closes the compose box.
+
+A common scenario is that a user is just pausing to think for a few
+seconds, but they still intend to finish the message.  Of course,
+that's hard to distinguish from the scenario of the user got pulled
+away from their desk.  For the former case, where the "writing user"
+completes the message with lots of pauses for thinking, a series of
+"start" and "stop" messages may be sent over time.  One way to
+improve our feature here is to tweak the timeout values, but of
+course it is always a tradeoff depending on how quickly people type,
+how long they pause to think, how frequently they get interrupted, etc.
+
+## Server
+
+The server piece of typing notificiations is currently pretty
+straightforward, since we take advantage of Zulip's event system,
+and the server piece here is basically a single Django view
+function with a small bit of library support to send out events
+to clients.
+
+Requests come into `/json/typing`.  The view mostly calls out
+to `check_send_typing_notification` to do the heavy lifting.
+
+One of the main things that the server does is to simply validate
+the recipients with a call to `recipient_for_emails`.  (We should
+streamline the payload in the request to have user ids instead of
+emails, but of course we will still need to validate recipients.)
+
+Once the request has been validated, the server sends events to
+potential recipients of the message.  The event type for that
+payload is `typing`.  See the function `do_send_typing_notification`
+for more details.
+
+## Reading user
+
+When a user plays the role of a "reading user," the client handles
+incoming "typing" events from the server, and the client will
+display typing notification only when both of these conditions are
+true:
+
+* The "writing user" is still likely typing.
+* The "reading user" is in a view where they'd see the eventual
+  message.
+
+The simplest case to consider is that the "reading user" is narrowed
+to a 1:1 conversation with the "writing user."  As the typing
+notification events come in to the client, a status message is shown
+just below the message feed.  It's possible for the "reading user"
+to be in a broader PM view, too, and when that happens, we look
+at the incoming events for all potential users that can be typing
+messages to us.
+
+The events will come in to `static/js/server_events_dispatch.js`.
+The `stop` and `start` operations get further handled by
+`static/js/typing_events.js`.
+
+The client maintains a list of incoming "typists" using code in
+`static/js/typing_data.js`.
+
+One subtle thing that the client has to do here is to maintain
+timers for typing notifications.  The constant
+`TYPING_STARTED_EXPIRY_PERIOD` is used to determine that the
+"writing user" has abandoned the message.  Of course, the
+"writing user" will also explicitly send us `stop` notifications
+at certain times.
```

Comment:
I think somewhere, we should mention the required relationships between the various time periods (e.g. expiry timeout > retransmit frequency + maximum network delay), and why we need these timers (the reason is that we need to handle the case where a client disappears from the Internet without letting us know the user has stopped typing)